File: JavaScript/221-Maximal Square.js
/**
 * @param {character[][]} matrix
 * @return {number}
 */
var maximalSquare = function(matrix) {
    let row = matrix.length, col = matrix[0].length;
    let res = 0;
    
    
    for (let r=0; r<row; r++){
        for (let c=0; c<col; c++){
            if (matrix[r][c] === "0"){
                matrix[r][c] = 0;
            }else{
                if (r-1>=0 && c-1 >= 0){
                    matrix[r][c] = Math.min(matrix[r-1][c-1], matrix[r][c-1], matrix[r-1][c])+1;
                }else{
                    matrix[r][c] = parseInt(matrix[r][c]);
                }
                res = Math.max(matrix[r][c], res);
            }
        }
    }
    
    return res*res;
};

File: 221-Maximal Square.py
class Solution(object):
    def maximalSquare(self, matrix):
        """
        :type matrix: List[List[str]]
        :rtype: int
        """
        
# dp(i,j) = min(dp(i−1,j),dp(i−1,j−1),dp(i,j−1))+1.

        row = len(matrix)
        col = len(matrix[0])
        
        res = 0
        
        for r in range(row):
            for c in range(col):
                if matrix[r][c] != "0":
                    if r-1 >= 0 and c-1 >= 0:
                        matrix[r][c] = min(matrix[r-1][c], matrix[r][c-1], matrix[r-1][c-1])+1
                    else:
                        matrix[r][c] = int(matrix[r][c])
                    res = max(res, int(matrix[r][c]) )
                else:
                    matrix[r][c] = 0
                
        return res*res